### DISCUSSION TITLE
Rendering an SVG to given pixel dimensions: should `-density` always be one and a half times the desired dimensions?

### DISCUSSION THREAD

**Discussion by peteroupc** [ImageMagick/ImageMagick#7440] [Help] [answered]
### ImageMagick version

7.1.1-26

### Operating system, version and so on

Fedora 40

### Description

Suppose I want to render an SVG with a particular width and height _in pixels_, disregarding the `width` and `height` attributes given on the SVG.

Then, at least if the desired width and height are the same, I can do so with the following ImageMagick command:

    magick -density wxh image.svg image.png

Where `w` is one and a half times the desired width in pixels, and `h` is one and a half times the desired height in pixels.

But, given that density is dots per inch by default, which may vary by computer, are `w` and `h` given above _always_ one and a half times the desired dimensions in pixels?

**Comment by fmw42** [comment]
Why not just edit the width and height in the SVG file?

**Reply by peteroupc** [reply]
ImageMagick will be used programmatically to render an arbitrary SVG file at a given width and height in pixels.  Editing the width and height of an arbitrary SVG file programmatically will require using an XML parser on the SVG file, editing the SVG's width and height attributes, and saving a temporary copy of the SVG to disk; this programmatic process is much too heavyweight for my purposes, especially if a program like ImageMagick could render an SVG to user-defined pixel dimensions.

**Comment by snibgo** [comment]
> should -density always be one and a half times the desired dimensions?

No. That would be true only for SVGs that have dimensions of 2/3 (0.6667) of an inch.

Instead of `-density`, you can use `-size WxH` before reading the SVG.

**Reply by peteroupc** [reply]
Take the following SVG, with a given width and height of 64&times;64.

```
<svg width="64" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 64 64" height="64" xmlns:xlink="http://www.w3.org/1999/xlink">
  <ellipse style='stroke:none;fill:blue' cx='32' cy='32' rx='22' ry='22'/>
</svg>
```

Then the `-size` command won't scale the SVG, as opposed to `density`.  That is, the command:

    magick -size 256x256 image.svg image.png
    
...won't scale the SVG to 256&times;256 upon rendering it to a PNG.

Meanwhile, the command:

    magick -density 384x384 image.svg image.png
    
...will scale the SVG to 256&times;256.

**Comment by snibgo** [comment]
Your SVG has internal dimensions of 64x64. At the default SVG density of 96 dpi, this is 64/96=0.6667 inches. This is why, for this SVG, you can use a density of one and a half times the desired dimensions.

But it won't work for SVGs of other internal dimensions. For example, if you change the SVG file so every "64" becomes "640".

"-size" works for me, with IM v7.1.1-20. For example, where `svg64.svg` is your file, and `svg640.svg` is your file after changing every `64` to `640`:
```
C:\web\im>%IMG7%magick -size 200x200 svg64.svg info:
svg64.svg SVG 200x200 200x200+0+0 16-bit sRGB 210B 0.000u 0:00.000

C:\web\im>%IMG7%magick -size 200x200 svg640.svg info:
svg640.svg SVG 200x200 200x200+0+0 16-bit sRGB 214B 0.000u 0:00.000
```

**Reply by peteroupc** [reply]
That doesn't appear to be the case with version 7.1.1-33:


```
****$ magick -version
Version: ImageMagick 7.1.1-33 Q16-HDRI x86_64 22263 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): bzlib cairo djvu fftw fontconfig freetype gslib gvc heic jbig jng jp2 jpeg jxl lcms lqr ltdl lzma openexr pangocairo png ps raqm raw rsvg tiff webp wmf x xml zip zlib zstd
Compiler: gcc (14.1)
****$ magick -size 200x200 svg64.svg info:
/tmp/atom.svg SVG 64x64 64x64+0+0 8-bit TrueColorAlpha sRGB 909B 0.000u 0:00.000
```

Here, the desired size is somehow ignored.

**Reply by dlemstra** [reply]
The desires size won't be used when you are using inkscape. Are you using that?

**Reply by peteroupc** [reply]
> The desires size won't be used when you are using inkscape. Are you using that?

No.

**Reply by peteroupc** [reply]
Can you investigate the apparent problem with `-size`  in ImageMagick 7.1.1-33?

**Reply by dlemstra** [reply, marked as answer]
Did you check if `inkscape` was installed on your system? It will be used when it is available. For both `msvg` and `rsvg` the size is used:

```
>magick -size 256x256 mvsg:image.svg info:
mvsg:image.svg SVG 256x256 256x256+0+0 16-bit sRGB 210B 0.001u 0:00.000

>magick -size 256x256 rvsg:image.svg info:
rvsg:image.svg SVG 256x256 256x256+0+0 16-bit sRGB 210B 0.001u 0:00.000
```

**Reply by peteroupc** [reply]
Thank you for responding.  Both `msvg:` and `rsvg:` (not: `mvsg:` and `rvsg:`)  take account of `-size` and render the SVG at the given size.  There are interesting differences I noticed though:

- `msvg:` seems to flatten curved paths into straight lines (assuming the SVG will be displayed at its original size) before rendering the SVG at the desired pixel size.
- `rsvg:` renders curved paths more smoothly because the flattening takes place at the desired pixel size, not the original size.  For this reason, `rsvg:` is more suitable for my purposes.

I think these points should be mentioned somewhere in the documentation (or `msvg:` improved to render like `rsvg:` here), as well as that ImageMagick uses Inkscape to render SVGs by default when that program is installed and Inkscape doesn't take into account the `-size` option.

**Reply by peteroupc** [reply]
I have [filed an issue](https://github.com/ImageMagick/ImageMagick/issues/7465) about `msvg:` flattening curved paths too coarsely.

Now, as for `-size` being potentially ignored depending on whether Inkscape is installed: What should be the appropriate behavior for the following two ImageMagick commands?

    magick -size 256x256 vector.svg raster.png
    magick vector.svg raster.png
    
In my opinion, if `-size` is given, ImageMagick should use by default an SVG renderer that takes account of `-size` (if not Inkscape, then possibly `rsvg`).  If `-size` is not given, then ImageMagick can safely default to Inkscape.  I find it surprising that `-size` can be ignored or not depending on whether Inkscape is installed, especially since that behavior is not mentioned in ImageMagick's documentation as far as I can tell. No matter how the question above is answered, the answer should be mentioned in the documentation.

**Reply by fmw42** [reply]
Apparently it is an Inkscape limitation, not under the control of Imagemagick.
